### PR TITLE
TW-1695: Leave group

### PR DESCRIPTION
**Pull request by hieutbui**
## Ticket
- #1695 

## Resolved
- Web:

https://github.com/linagora/twake-on-matrix/assets/80142234/5ab97265-7ccb-49b3-bcbe-9ea44d48c9ab

- Android:

https://github.com/linagora/twake-on-matrix/assets/80142234/78f15013-b2b7-49d3-b75a-fb31363af97b

### CODE REVIEW

File: lib/pages/chat_details/chat_details_edit.dart [hunk 1]
```diff
@@ -494,6 +495,22 @@ class ChatDetailsEditController extends State<ChatDetailsEdit>
     });
   }
 
+  Future<void> leaveChat() async {
+    final currentRoom = room;
+    if (currentRoom == null) {
+      throw Exception(
+        'Leave room button clicked while room is null. This should not be possible from the UI!',
+      );
+    }
```
Comment (sherlockvn) [inline lib/pages/chat_details/chat_details_edit.dart:504]: you throw the exception, but i don't see where you handle it

Comment (sherlockvn) [reply lib/pages/chat_details/chat_details_edit.dart:504]: i think you should wrap the method in try catch

File: lib/pages/chat_details/chat_details_edit_option.dart [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+                        children: [
+                          Text(
+                            title,
+                            style: Theme.of(context)
+                                .textTheme
+                                .titleMedium
+                                ?.copyWith(
+                                  color: titleColor ??
+                                      Theme.of(context).colorScheme.onSurface,
+                                  fontSize: 16.0,
```
Comment (sherlockvn) [inline lib/pages/chat_details/chat_details_edit_option.dart:72]: to the style also

File: lib/pages/chat_details/chat_details_edit_option.dart [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+                            child: Text(
+                              subtitle,
+                              style: Theme.of(context)
+                                  .textTheme
+                                  .bodySmall
+                                  ?.copyWith(
+                                    color: subtitleColor ??
+                                        LinagoraRefColors.material()
+                                            .neutral[40],
+                                    fontSize: 12.0,
```
Comment (sherlockvn) [inline lib/pages/chat_details/chat_details_edit_option.dart:89]: same

File: lib/pages/chat_details/chat_details_edit_option.dart (hunk 2 above)
Comment (hieutbui) [reply lib/pages/chat_details/chat_details_edit_option.dart]: done

File: lib/pages/chat_details/chat_details_edit_option.dart (hunk 3 above)
Comment (hieutbui) [reply lib/pages/chat_details/chat_details_edit_option.dart]: done

**Comment by sherlockvn** [comment]
Fix this bug plz:

https://github.com/linagora/twake-on-matrix/assets/43041967/abce6efa-c487-40e9-9754-e9a1e930d663

**Comment by hoangdat** [comment]
- [ ] constraint width for context menu
- [ ] should hide context menu when not use

File: lib/pages/chat_details/chat_details_edit_view.dart [hunk 4]
```diff
@@ -179,8 +183,28 @@ class ChatDetailsEditView extends StatelessWidget {
...
+            ),
+            Container(
+              color: Theme.of(context).colorScheme.surfaceVariant,
+              alignment: Alignment.centerLeft,
+              padding: const EdgeInsets.all(8.0),
+              child: Text(
+                L10n.of(context)!.dangerZone,
+                style: Theme.of(context).textTheme.titleSmall?.copyWith(
+                      color: Theme.of(context).colorScheme.tertiary,
+                      fontSize: 14.0,
```
[the remark this answers is not in the document]
Comment (hieutbui) [reply lib/pages/chat_details/chat_details_edit_view.dart]: done

**Comment by hieutbui** [comment]
> Fix this bug plz:
> 
>  Screen.Recording.2024-04-23.at.09.19.15.mov

fixed!

**Review by sherlockvn — APPROVED** [review]

File: lib/pages/chat/chat.dart [hunk 5]
```diff
@@ -311,17 +313,27 @@ class ChatController extends State<Chat>
...
-        'Leave room button clicked while room is null. This should not be possible from the UI!',
+    try {
+      final room = this.room;
+      if (room == null) {
+        throw RoomNullException();
+      }
+
+      final success = await TwakeDialog.showFutureLoadingDialogFullScreen(
+        future: room.leave,
+      );
```
[the remark this answers is not in the document]
Comment (hieutbui) [reply lib/pages/chat/chat.dart]: done

File: lib/pages/chat/chat.dart [hunk 6]
```diff
@@ -311,17 +313,27 @@ class ChatController extends State<Chat>
...
+
+      if (result.error != null) return;
+      context.go('/rooms');
+    } on RoomNullException catch (e) {
+      Logs().e(
+        'Chat::leaveChat() - RoomNullException - $e',
+      );
+    } catch (e) {
+      Logs().e(
```
[the remark this answers is not in the document]
Comment (hieutbui) [reply lib/pages/chat/chat.dart]: I'll show the `Sneak bar` for error case

File: lib/pages/chat_details/chat_details_edit.dart [hunk 7]
```diff
@@ -494,6 +496,30 @@ class ChatDetailsEditController extends State<ChatDetailsEdit>
     });
   }
 
+  Future<void> leaveChat() async {
```
[the remark this answers is not in the document]
Comment (hieutbui) [reply lib/pages/chat_details/chat_details_edit.dart]: Separated!

File: lib/pages/chat/chat.dart [hunk 8]
```diff
@@ -1776,39 +1765,48 @@ class ChatController extends State<Chat>
...
+  ) {
+    final offset = tapDownDetails.globalPosition;
+    showTwakeContextMenu(
+      offset: offset,
+      context: context,
+      builder: (_) => _appbarMenuActionTile(context),
+    );
+  }
+
+  List<Widget> _appbarMenuActionTile(BuildContext context) {
```
Comment (hoangdat) [inline lib/pages/chat/chat.dart:1780]: please separate app bar for: select mode and normal mode

**Review by hoangdat — APPROVED** [review]

**Review by Te-Z — APPROVED** [review]

File: lib/pages/chat/chat.dart (hunk 8 above)
Comment (hieutbui) [reply lib/pages/chat/chat.dart]: done

**Review by hoangdat — APPROVED** [review]

**Pull request merged**